### ISSUE TITLE
Error when using lazy attribute with link resource

### ISSUE THREAD

**Issue by jimcm** [chef/chef#3703]
### Version:

```
/usr/bin/chef-client -version
Chef: 12.1.2
```
### Environment: [Details about the environment such as the Operating System, cookbook details, etc.]

Red Hat Enterprise Linux Server release 6.6 (Santiago)
### Scenario:

[What you are trying to achieve and you can't?]

Use lazy attribute evaluation for target_file attribute of the link resource. Getting an error,
`"No implicit conversion of Chef::DelayedEvaluator into String"`
### Steps to Reproduce:

[If you are filing an issue what are the things we need to do in order to repro your problem?]

Details from the chef client run, including the recipe/resource declaration:

```
$ /usr/bin/chef-client
Starting Chef Client, version 12.1.2
resolving cookbooks for run list: ["testcb::default"]
Synchronizing Cookbooks:
  - testcb
Compiling Cookbooks...
Converging 1 resources
Recipe: testcb::link
  * link[test-resource] action create


    Error executing action `create` on resource 'link[test-resource]'


    TypeError

    no implicit conversion of Chef::DelayedEvaluator into String

    Resource Declaration:

     In /var/chef/cache/cookbooks/testcb/recipes/link.rb

      2: link 'test-resource' do
      3:   target_file lazy { '/tmp/hardlink' }
      4:   to '/tmp/testfile'
      5:   link_type :hard
      6: end

    Compiled Resource:

     Declared in /var/chef/cache/cookbooks/testcb/recipes/link.rb:2:in `from_file'

    link("test-resource") do
      action :create
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      to "/tmp/testfile"
      link_type :hard
      target_file #<Chef::DelayedEvaluator:0x00000004059fe8@/var/chef/cache/cookbooks/testcb/recipes/link.rb:3>
      declared_type :link
      cookbook_name "testcb"
      recipe_name "link"
    end


Running handlers:
Running handlers complete
Chef Client failed. 0 resources updated in 1.807145476 seconds
```

**Comment by lamont-granquist** [comment]
This is probably reasonably easy to fix, just another of the endless whack-a-mole of fixing lazy attributes...

**Comment by ostsjoe** [comment]
This appears to still be happening in at least 12.13 an 12.14 when "end.run_action(:create)" is used to execute a resource during the compile phase. See reference above.

**Comment by tay** [comment]
ran into this issue with chef 11 but looks it was resolved downstream https://github.com/chef/chef/pull/3130 ?

**Comment by thommay** [comment]
This was fixed in 12.3
